### PR TITLE
New Rule Proj1001: Use analyzers for packages

### PR DESCRIPTION
**Pull request by Corniel**
A rule that should alert when some useful analyzers (based on the project dependencies) ar missing.

-----------------
# Proj1001: Use analyzers for packages
Some [NuGet](https://www.nuget.org/) packages come with there own/dedicated
Roslyn analyzers; they just contain rules to improve the usage of those packages.
In order to get the best out of those NuGet packages, their analyzer(s) should
be used.

## Analyzers for packages
The following analyzers are known to be beneficial for some packages:

| Analyzer(s)                                                                                                           | Packages                                                                                            |
|-----------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------|
| [FluentAssertions.Analyzers](https://www.nuget.org/packages/FluentAssertions.Analyzers)                               | [FluentAssertions.*](https://www.nuget.org/packages?q=FluentAssertions.*)                           |
| [Microsoft.AspNetCore.Components.Analyzers](https://www.nuget.org/packages/Microsoft.AspNetCore.Components.Analyzers) | [Microsoft.AspNetCore.*](https://www.nuget.org/packages?q=Microsoft.AspNetCore.)                    |
| [Microsoft.Azure.Functions.Analyzers](https://www.nuget.org/packages/Microsoft.Azure.Functions.Analyzers)             | [Microsoft.Azure.Functions.*](https://www.nuget.org/packages?q=Microsoft.Azure.Functions.*)         |
| [Microsoft.EntityFrameworkCore.Analyzers](https://www.nuget.org/packages/Microsoft.EntityFrameworkCore.Analyzers)     | [Microsoft.EntityFrameworkCore.*](https://www.nuget.org/packages?q=Microsoft.EntityFrameworkCore.*) |
| [MongoDB.Analyzer](https://www.nuget.org/packages/MongoDB.Analyzer)                                                   | [MongoDB.*](https://www.nuget.org/packages?q=MongoDB.*)                                             |
| [NUnit.Analyzers](https://www.nuget.org/packages/NUnit.Analyzers)                                                     | [NUnit.*](https://www.nuget.org/packages?q=NUnit.*)                                                 |
| [Qowaiv.Analyzers.CSharp](https://www.nuget.org/packages/Qowaiv.Analyzers.CSharp)                                     | [Qowaiv.*](https://www.nuget.org/packages?q=Qowaiv.*)                                               |
| [SerilogAnalyzer](https://www.nuget.org/packages/SerilogAnalyzer)                                                     | [Serilog.*](https://www.nuget.org/packages?q=Serilog.*)                                             |
| [xunit.analyzers](https://www.nuget.org/packages/xunit.analyzers)                                                     | [xunit.*](https://www.nuget.org/packages?q=xunit.*)                                                 |

### CODE REVIEW

File: src/DotNetProjectFile.Analyzers/Analyzers/UseAnalyzersForPackages.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            if (packageReferences.None(analyzer.IsMatch)
+                && context.Compilation.ReferencedAssemblyNames.FirstOrDefault(analyzer.IsMatch) is { } reference)
+            {
+                context.ReportDiagnostic(Descriptor, analyzer.Package, reference.Name);
+            }
+        }
+    }
+
+    public static readonly PackageAnalyzer[] Analyzers = new PackageAnalyzer[]
```
Comment (ROSDEV) [inline src/DotNetProjectFile.Analyzers/Analyzers/UseAnalyzersForPackages.cs:29]: See github warning

File: dotnet-project-file-analyzers.sln [hunk 2]
```diff
@@ -22,6 +22,7 @@ EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Rules", "Rules", "{50507CD3-2049-4539-8A37-D370EE0F91F7}"
 	ProjectSection(SolutionItems) = preProject
 		rules\Proj0001.md = rules\Proj0001.md
+		rules\Proj1001.md = rules\Proj1001.md
```
Comment (LauraXiulan) [inline dotnet-project-file-analyzers.sln:25]: Duplicate

File: src/DotNetProjectFile.Analyzers/Analyzers/UseAnalyzersForPackages.cs (hunk 1 above)
Comment (LauraXiulan) [reply src/DotNetProjectFile.Analyzers/Analyzers/UseAnalyzersForPackages.cs:29]: False positive.

File: src/DotNetProjectFile.Analyzers/Analyzers/UseAnalyzersForPackages.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        new("Microsoft.Azure.Functions.Analyzers", "Microsoft.Azure.Functions"),
+        new("Microsoft.EntityFrameworkCore.Analyzers", "Microsoft.EntityFrameworkCore"),
+        new("MongoDB.Analyzer", "MongoDB"),
+        new("NUnit.Analyzers", "NUnit"),
+        new("Qowaiv.Analyzers.CSharp", "Qowaiv"),
+        new("SerilogAnalyzer", "Serilog"),
+        new("xunit.analyzers", "xunit"),
+    };
+
+    public sealed record PackageAnalyzer
```
Comment (LauraXiulan) [inline src/DotNetProjectFile.Analyzers/Analyzers/UseAnalyzersForPackages.cs:42]: private

File: specs/DotNetProjectFile.Analyzers.Specs/FluentAssertions/FindsExtensions.cs [hunk 4]
```diff
@@ -18,6 +18,11 @@ public static void HasIssues(this ProjectAnalyzerVerifyContext context, params I
             .Select(Issue.FromDiagnostic)
             .ToArray();
 
+        foreach(var issue in reported) 
+        {
```
[the remark this answers is not in the document]
Comment (Corniel) [reply specs/DotNetProjectFile.Analyzers.Specs/FluentAssertions/FindsExtensions.cs:22]: Not fully.

**Pull request merged**